src/gui/raster/qgssinglebandpseudocolorrendererwidget.cpp

```diff
@@ -157,30 +192,58 @@ QgsRasterRenderer* QgsSingleBandPseudoColorRendererWidget::renderer()
   return renderer;
 }
 
-void QgsSingleBandPseudoColorRendererWidget::on_mAddEntryButton_clicked()
-{
-  QTreeWidgetItem* newItem = new QTreeWidgetItem( mColormapTreeWidget );
-  newItem->setText( 0, "0.0" );
-  newItem->setBackground( 1, QBrush( QColor( Qt::magenta ) ) );
-  newItem->setText( 2, tr( "Custom color map entry" ) );
-}
-
-void QgsSingleBandPseudoColorRendererWidget::on_mDeleteEntryButton_clicked()
+/** Generate labels from the values in the color map.
+ *  Skip labels which were manually edited (black text).
+ *  Text of generated labels is made gray
+ */
+void QgsSingleBandPseudoColorRendererWidget::autoLabel()
 {
-  QTreeWidgetItem* currentItem = mColormapTreeWidget->currentItem();
-  if ( currentItem )
+  bool discrete = mColorInterpolationComboBox->currentText() == tr( "Discrete" );
+  QString unit = mUnitLineEdit->text();
+  QString label = "";
+  int myTopLevelItemCount = mColormapTreeWidget->topLevelItemCount();
+  QTreeWidgetItem* myCurrentItem;
+  for ( int i = 0; i < myTopLevelItemCount; ++i )
   {
-    delete currentItem;
+    myCurrentItem = mColormapTreeWidget->topLevelItem( i );
+    //If the item is null or does not have a pixel values set, skip
+    if ( !myCurrentItem || myCurrentItem->text( 0 ) == "" )
+    {
+      continue;
+    }
+
+    if ( discrete )
+    {
+      if ( i == 0 )
+      {
+        label = "< " + myCurrentItem->text( 0 ) + unit;
+      }
+      else
+      {
+        label = mColormapTreeWidget->topLevelItem( i - 1 )->text( 0 ) + " - " + myCurrentItem->text( 0 ) + unit;
+      }
+    }
+    else
+    {
+      label = myCurrentItem->text( 0 ) + unit;
+    }
+
+    if ( myCurrentItem->text( 2 ) == "" || myCurrentItem->text( 2 ) == label || myCurrentItem->foreground( 2 ).color() == QColor( Qt::gray ) )
+    {
+      myCurrentItem->setText( 2, label );
+      myCurrentItem->setForeground( 2, QBrush( QColor( Qt::gray ) ) );
+    }
   }
 }
 
-void QgsSingleBandPseudoColorRendererWidget::on_mSortButton_clicked()
+/** Extract the unit out of the current labels and set the unit field. */
+void QgsSingleBandPseudoColorRendererWidget::setUnitFromLabels()
 {
-  bool inserted = false;
-  int myCurrentIndex = 0;
+  bool discrete = mColorInterpolationComboBox->currentText() == tr( "Discrete" );
+  QStringList allSuffixes;
+  QString label = "";
```

Comment:
Just go `QString label` - no need to initialize to an empty string, that's already done by QString constructor.